Thermostat preset with modes

MrMDavidson opened the pull request:
# What does this implement/fix?

This introduces the ability to configure additional presets for a Thermostat beyond the existing Home and Away modes. The existing configuration will continue to work. But it also introduces a new configuration list for preset modes. New style presets also support default fan, swing, and operating modes. Additionally a new trigger has been introduced - `preset_change` - which fires when a user has changed the current preset; intended to be used to update visual displays in the same style as the `auto_mode`, etc, triggers. 

The existing Home/Away logic has been consolidated with this generic preset approach.

Apologies for any faux pas in the Python side of it - it's not my forte.

This supersedes #3266 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1966

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
climate:
  - platform: thermostat
    # Kept for backwards compatability
    away_config:
      default_target_temperature_low: 16
      default_target_temperature_high: 26

    preset:
      activity:
        default_target_temperature_low: 21
        default_target_temperature_high: 23
        fan_mode high
        swing_mode: both
      sleep:
        default_target_temperature_low: 17
        default_target_temperature_high: 26
        fan_mode: low
        mode: fan_only
      # etc. Can be any of eco, away, boost, comfort, home, sleep, and activity
    preset_change:
      - logger.log: Preset has been changed
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

MrMDavidson committed:
Rework HOME/AWAY support to being driven via a map of ClimatePreset/ThermostatClimateTargetTempConfig
This opens up to theoretically being able to support other presets (ECO, SLEEP, etc)

MrMDavidson committed:
Add support for additional presets
Configuration takes the form;
```
climate:
  platform: preset
  ...
  preset:
    [eco | away | boost | comfort | home | sleep | activity]:
      default_target_temperature_low: 20
      default_target_temperature_high: 24
```

These will be available in the Home Assistant UI and, like the existing Home/Away config will reset the temperature in line with these defaults when selected. The existing away_config/home_config is still respected (although preset->home/preset->away will be applied after them and override them if both styles are specified)

MrMDavidson committed:
Add support for specifying MODE, FAN_MODE and SWING_MODE on a preset
When switching presets these will implicitly flow through to the controller. However calls to climate.control which specify any of these will take precedence even when changing the mode (think of the preset version as the default for that preset)

MrMDavidson committed:
Add `preset_change` mode trigger
When defined this trigger will fire when the preset for the thermostat has been changed. The intent of this is similar to `auto_mode` - it's not intended to be used to control the preset's state (eg. communicate with the physical thermostat) but instead might be used to update a visual indicator, for instance.

MrMDavidson committed:
Apply lint, clang-format, and clang-tidy fixes

MrMDavidson committed:
Additional clang-format fixes

MrMDavidson committed:
Wrap log related strings in LOG_STR_ARG

kbx81 commented:
I'm wondering if, instead of this arrangement:
```yaml
    preset:
      activity:
        default_target_temperature_low: 21
        default_target_temperature_high: 23
        fan_mode high
        swing_mode: both
      sleep:
        default_target_temperature_low: 17
        default_target_temperature_high: 26
        fan_mode: low
        mode: fan_only
```
...it would make more sense to do this:
```yaml
    preset:
      - name: "activity"
        default_target_temperature_low: 21
        default_target_temperature_high: 23
        fan_mode: high
        swing_mode: both
      - name: "sleep"
        default_target_temperature_low: 17
        default_target_temperature_high: 26
        fan_mode: low
        mode: fan_only
```

I feel like this would provide a more consistent way to define presets with custom names. Any thoughts?

MrMDavidson commented:
> I'm wondering if, instead of this arrangement:
> 
> ```yaml
>     preset:
>       activity:
>         default_target_temperature_low: 21
>         default_target_temperature_high: 23
>         fan_mode high
>         swing_mode: both
>       sleep:
>         default_target_temperature_low: 17
>         default_target_temperature_high: 26
>         fan_mode: low
>         mode: fan_only
> ```
> 
> ...it would make more sense to do this:
> 
> ```yaml
>     preset:
>       - name: "activity"
>         default_target_temperature_low: 21
>         default_target_temperature_high: 23
>         fan_mode: high
>         swing_mode: both
>       - name: "sleep"
>         default_target_temperature_low: 17
>         default_target_temperature_high: 26
>         fan_mode: low
>         mode: fan_only
> ```
> 
> I feel like this would provide a more consistent way to define presets with custom names. Any thoughts?

I'm definitely a fan of moving the type to a child node rather than having it as the parent as I think it makes it easier to document (I found the current way a bit difficult to articulate without an example)

The Name there might be a bit overloaded, though. Wonder if it might be worthwhile using a required "type" and a "name" (Required for customs, can be used just as a debug label for others). 

```yaml
preset:
  - type: activity
    ...
  - type: custom
    name: "My custom activity..."
```

As we've missed the 2022.03 release I'm happy to have more of a play with this and try and get custom presets into this same PR.

kbx81 commented:
Sweet! I'm on board -- I say have a go at that approach. I think it would be cleaner overall, but yes, I also agree about the docs (I was thinking the same thing 😆).

I'm testing/playing around with it as you have it now and it's working for me! Definitely a nice feature to (finally) have -- again, thank you so much for working on this! 😄🍻

kbx81 commented:
> Wonder if it might be worthwhile using a required "type" and a "name" 

Another thought on this might be to just use `name` and, if the name doesn't match one of the predefined names, it is assumed to be of type `custom` and so we set it up/register it that way.

MrMDavidson commented:
> Another approach might be to just use `name` and, if the name doesn't match one of the predefined names, it is assumed to be of type `custom` and so we set it up/register it that way

One reason I'd err against this is that if you typo and spell it "selep" instead of "sleep" you'd end up with a custom preset called "selep" vs a validation error. I'm not sure what flow on effects this would cause; I'm not sure if the custom vs built in presets are handled differently in other parts of the code base (I do know that `climate.control` has a `preset` and `custom_preset` field)

kbx81 commented:
I don't know enough about it to have a strong opinion -- but I will say that I feel like it's a bit odd that we have a block of "built-in" presets and then this other strategy for defining "custom" presets. It seems like unnecessary complication to me...but, again, I don't know much about the inner workings so perhaps there is a reason for distinguishing them in such a manner.

In terms of documentation and minimizing cognitive load, my perspective is that it's best to keep it as simple as possible in terms of how it is set up in the yaml config...hence my original suggestion. But that's just my take -- I'm sure others will have different perspectives. Hopefully some others will chime in here!

I'm not sure I'd stress too much about something like a typo in the config -- I agree that it would be nice to have the validation catch that but at the same time, there are plenty of other similar "gotchas" in the config that would trip someone up in just the same way. For example, if I name a `switch` or a `light` incorrectly, the validation won't help me there and I'd end up on the same goose chase. 🙃

kbx81 commented:
> I do know that climate.control has a preset and custom_preset field

Ah...in that case, it would probably would be best to break them apart. Maybe something like this?:
```yaml
    preset:
      - type: builtin
        name: "boost"
        default_target_temperature_low: 70°F
        default_target_temperature_high: 72°F
      - type: builtin
        name: "comfort"
        default_target_temperature_low: 74°F
        default_target_temperature_high: 76°F
      - type: custom
        name: "workout"
        default_target_temperature_low: 68°F
        default_target_temperature_high: 78°F
```
Again, I don't have a strong opinion here...I'm just throwing out ideas. 😄

MrMDavidson commented:
> I don't know enough about it to have a strong opinion -- but I will say that I feel like it's a bit odd that we have a block of "built-in" presets and then this other strategy for defining "custom" presets. It seems like unnecessary complication to me...but, again, I don't know much about the inner workings so perhaps there is a reason for distinguishing them in such a manner.

Definitely agree - the implementation I'm planning here they'd all get lumped in the same bag for the Thermostat implementation, so it shouldn't matter for us. It's just if HA / the base Climate cares for some reason.

> In terms of documentation and minimizing cognitive load, my perspective is that it's best to keep it as simple as possible in terms of how it is set up in the yaml config...hence my original suggestion. 

+1 for this. There's more people making use of the config than there are implementing the code for it :)

> Ah...in that case, it would probably would be best to break them apart. Maybe something like this?:
> ```yaml
>     preset:
>       - type: builtin
>         name: "boost"
>         default_target_temperature_low: 70°F
>         default_target_temperature_high: 72°F
>       - type: builtin
>         name: "comfort"
>         default_target_temperature_low: 74°F
>         default_target_temperature_high: 76°F
>       - type: custom
>         name: "workout"
>         default_target_temperature_low: 68°F
>         default_target_temperature_high: 78°F
> ```

I think if there does prove to be a difference between the two; this is a good approach to the config. I like it.

MrMDavidson committed:
Add support for custom presets
This also changes the configuration syntax to;
```yaml
  preset:
    # Standard preset
    - name: [eco | away | boost | comfort | home | sleep | activity]
      default_target_temperature_low: 20
      ...
    # Custom preset
    - name: My custom preset
      default_target_temperature_low: 18
```

For the end user there is no difference between a custom and built in preset. For developers custom presets are set via `climate.control` `custom_preset` property instead of the `preset`

MrMDavidson commented:
> I'm wondering if, instead of this arrangement:
> 
> ```yaml
>     preset:
>       activity:
>         default_target_temperature_low: 21
>         default_target_temperature_high: 23
>         fan_mode high
>         swing_mode: both
>       sleep:
>         default_target_temperature_low: 17
>         default_target_temperature_high: 26
>         fan_mode: low
>         mode: fan_only
> ```
> 
> ...it would make more sense to do this:
> 
> ```yaml
>     preset:
>       - name: "activity"
>         default_target_temperature_low: 21
>         default_target_temperature_high: 23
>         fan_mode: high
>         swing_mode: both
>       - name: "sleep"
>         default_target_temperature_low: 17
>         default_target_temperature_high: 26
>         fan_mode: low
>         mode: fan_only
> ```
> 
> I feel like this would provide a more consistent way to define presets with custom names. Any thoughts?

Ended up going with this approach and added support for custom presets.

MrMDavidson committed:
Lint/clang-format/clang-tidy fixes

MrMDavidson committed:
Additional lint/clang-format/clang-tidy fixes

MrMDavidson committed:
Clang-tidy changes

MrMDavidson commented:
Hey @kbx81 not sure who the best person for this is but in making use of this I've uncovered what I think is a bug in the base `Climate` implementation. Looking at the docs for `climate.control` (https://esphome.io/components/climate/index.html#climate-control-action) you can see that `away`, `custom_fan_mode`, and `custom_preset` are all indicated as being templatable in the docs. However if you look at the code - https://github.com/esphome/esphome/blob/dev/esphome/components/climate/__init__.py#L279 - you can see that all of these are not validating as templates. As such a call to `climate.control` like...
```yaml
  then:
    - climate.control:
        id: my_hvac_system
        custom_preset: !lambda |-
          return "My Custom Preset Name"
```

Will fail to generate generate with an error message
>>> 
          custom_preset: !lambda |-
            return "My Custom Preset Name" ; [source /config/esphome/.climate_control.yaml:100]

However simply tweaking the config (Eg. https://github.com/esphome/esphome/blob/dev/esphome/components/climate/__init__.py#L292 ) to

```python
        cv.Exclusive(CONF_CUSTOM_PRESET, "preset"): cv.templatable(cv.string_strict),
```

Isn't enough as you'll get an error about the type...
>  File "/esphome/esphome/automation.py", line 323, in build_action_list
>    action = await build_action(conf, templ, arg_type)
>  File "/esphome/esphome/automation.py", line 316, in build_action
>    ret = await builder(config, action_id, template_arg, args)
>  File "/config/esphome/source/esphome/esphome/components/climate/__init__.py", line 333, in climate_control_to_code
>    template_ = await cg.templatable(config[CONF_CUSTOM_PRESET], args, str)
>  File "/esphome/esphome/cpp_generator.py", line 695, in templatable
>    return await process_lambda(value, args, return_type=output_type)
>  File "/esphome/esphome/cpp_generator.py", line 669, in process_lambda
>    return LambdaExpression(parts, parameters, capture, return_type, location)
>  File "/esphome/esphome/cpp_generator.py", line 227, in __init__
>    self.return_type = safe_exp(return_type) if return_type is not None else None
>  File "/esphome/esphome/cpp_generator.py", line 381, in safe_exp
>    raise ValueError("Object is not an expression", obj)
>ValueError: ('Object is not an expression', <class 'str'>)

The reason seems to be that when the call to `climate.control` is generated these particular properties are generated using `str` when they need to be generating the C++ type instead. Eg.
```python
    if CONF_CUSTOM_PRESET in config:
        template_ = await cg.templatable(config[CONF_CUSTOM_PRESET], args, cg.std_string)
        cg.add(var.set_custom_preset(template_))
```

Which then gets translated into the correct C++ code;
```cpp
climate_controlaction->set_custom_preset([=]() -> std::string {
      #line 101 "/config/esphome/.climate_control.yaml"
      return "My Custom Preset Name" ;
  });
```

Am I missing something here, or is this a bug? Preference for me to open a new PR for this fix?

kbx81 commented:
> Am I missing something here, or is this a bug? Preference for me to open a new PR for this fix?

I'd open an issue to report it and if you feel up to it, yes, please do a PR to fix! 😄

nagyrobi commented:
Is it possible to use lambdas for `default_target_temperature*`?

        name: "boost"
        default_target_temperature_low: 70°F
        default_target_temperature_high: 72°F

My use case would be to have a separate settings section where these values could be user-configurable. Like:
```
number:
- platform: template
  name: Boost target low
  id: boost_target_low
  optimistic: true
  max_value: 23
  min_value: 20
  initial_value: 21
- platform: template
  name: Boost target high
  id: boost_target_high
  optimistic: true
  max_value: 25
  min_value: 22
  initial_value: 23
```
and then:
```
    preset:
      - name: "boost"
        default_target_temperature_low: !lambda 'return id(boost_target_low).state;'
        default_target_temperature_high: !lambda 'return id(boost_target_high).state;'
```

MrMDavidson commented:
> Is it possible to use lambdas for `default_target_temperature*`?
> 
> My use case would be to have a separate settings section where these values could be user-configurable. Like:
> 
> ```
> number:
> - platform: template
>   name: Boost target low
>   id: boost_target_low
>   optimistic: true
>   max_value: 23
>   min_value: 20
>   initial_value: 21
> - platform: template
>   name: Boost target high
>   id: boost_target_high
>   optimistic: true
>   max_value: 25
>   min_value: 22
>   initial_value: 23
> ```
> 
> and then:
> 
> ```
>     preset:
>       - name: "boost"
>         default_target_temperature_low: !lambda 'return id(boost_target_low).state;'
>         default_target_temperature_high: !lambda 'return id(boost_target_high).state;'
> ```

At this stage no; because these just define what happens when you switch to the preset. However because they're only "defaults" they can be used in conjunction with a call to `climate.control`

```yaml

  then:
    climate.control:
      id: your_hvac_unit
      preset: boost
      target_temperature_low: !lambda 'return id(boost_target_low).state;'
      target_temperature_high: !lambda 'return id(boost_target_high).state;'
```

Or, alternatively, you could do it based on a trigger of the preset changing (untested);
```yaml
  climate:
    id: your_hvac_unit
    ...
    preset_change:
      - lambda: |-
          if (strcmp(id(your_hvac_unit).preset, "boost") == 0) {
            auto call = id(your_hvac_unit).make_call();
            call.set_target_temperature_low(id(boost_target_low).state);
            call.set_target_termperature_high(id(boost_target_high).state);
            
            call.perform();
          }
```

Weissnix4711 commented:
@nagyrobi Preset defaults should be updatable simply be running the set_preset_config function. Maybe have a on_value trigger on the number component, and have it run that function.

Alternatively, here is my approach: https://github.com/Weissnix4711/esphome/commit/4a676b65b05b193bb32f2e071cb07491eddc09e1

Every time the setpoint is changed, it will update the currently active preset's temperature setpoint, so if you switch to another preset and switch back, it will have remembered the last setpoint, and use that.

kbx81 commented on esphome/components/thermostat/climate.py:
```diff
@@ -58,8 +59,14 @@
+    CONF_PRESET,
+    CONF_FAN_MODE,
+    CONF_SWING_MODE,
+    CONF_MODE,
```
Usually these constants get alphabetized. Surprised the linter didn't call this out. Could you sort them accordingly so it's consistent with the rest of the codebase?
...line 6, too 😄

kbx81 commented on esphome/components/thermostat/climate.py:
```diff
@@ -415,6 +438,10 @@ def validate_thermostat(config):
+            cv.Optional(CONF_PRESET): cv.ensure_list(PRESET_CONFIG_SCHEMA),
+            cv.Optional(CONF_PRESET_CHANGE): automation.validate_automation(
+                single=True
+            ),
```
We should add a function to validate each preset in the list -- for example, it should make sure each preset specifies target temps that are consistent with what's in the main part of the config. In other words, if it only supports heat, a validation error should be thrown if `default_target_temperature_high` is present in any preset config. If it supports heating _and_ cooling, both `default_target_temperature_high` and `default_target_temperature_low` must be specified for each preset and, similarly, a validation error is thrown if this is not so.

This should also be done for the various fan and swing modes. Another example: if `fan_only_mode` is **not** in the config, then fan modes cannot be specified in the preset. You probably get the idea. 😄

MrMDavidson replied:
> We should add a function to validate each preset in the list -- for example, it should make sure each preset specifies target temps that are consistent with what's in the main part of the config. In other words, if it only supports heat, a validation error should be thrown if `default_target_temperature_high` is present in any preset config. If it supports heating _and_ cooling, both `default_target_temperature_high` and `default_target_temperature_low` must be specified for each preset and, similarly, a validation error is thrown if this is not so.

Good call. I've implemented this for temperature and modes.

> This should also be done for the various fan and swing modes. Another example: if `fan_only_mode` is **not** in the config, then fan modes cannot be specified in the preset. You probably get the idea. 😄

I don't think this logic is currently implemented for the default/away configs? Does it make sense? I can still have a controlled fan option on my A/C even in the absence of using fan only? (eg. It'll generate cool/hot air _and_ blow at a high speed. As opposed to only blowing room temperature air if `fan_only_mode` was set?)

kbx81 replied:
Ah, sorry...I wasn't clear. I was using `fan_only_mode` as an example. What I was getting at is that a validation error should be thrown if, for example, a given fan mode is specified in the preset but no supported fan mode is configured. Same for swing modes -- if a swing mode is specified in a preset but no supported swing mode is configured, validation error. As a more specific example, if no `fan_mode_auto_action` is configured, then setting the fan mode to `CLIMATE_FAN_AUTO` as part of a preset should trigger a validation error. Hopefully that's a bit clearer. 😄

You're right that this logic is not currently implemented, but the original default/away configs did not support changing modes/fan modes/swing modes -- they only supported temperatures so it wasn't needed. 😇

kbx81 replied:
Another point worth considering/validating is the range of temperatures that can be specified as part of a preset. As it is now, the visual min/max for the climate component is used to validate any given set point the user tries to set -- and if memory serves me there is internal validation the component does every time it recomputes its state. If a set point falls outside the bounds of the visual min/max, the set point is automatically reset to the min/max (whichever makes sense) to ensure reliable operation. If a preset specifies a set point outside of this range and it gets loaded as a set point, the internal validation will throw it away in favor of the appropriate visual min/max. This will undoubtedly confuse _someone_ as some point, so it would probably be worth throwing a validation error if, for example, I try to configure a preset with 0 degrees as a set point when the visual min/max range only allows (for example) 15 - 25. (You'll probably realize that, yes, this logic also did not yet exist. 😁)

MrMDavidson replied:
Ahh, all good! Have implemented `fan_mode`, `swing_mode`, `mode`, `default_target_temperature_low`, and `default_target_temperature_high` validation.

kbx81 commented on esphome/components/thermostat/thermostat_climate.cpp:
```diff
@@ -32,7 +32,7 @@ void ThermostatClimate::setup() {
     // restore from defaults, change_away handles temps for us
     this->mode = this->default_mode_;
-    this->change_away_(false);
+    this->change_preset_(climate::CLIMATE_PRESET_HOME);
```
Hmmm, I wonder if we should allow the default preset to be specified in the config... 🤔

MrMDavidson replied:
I'd kind of planned on leaving the `default*` and `away` config that currently exists (in main) as is, get the preset changes in, mark the existing config as deprecated and then in +1 releases time remove them. Thoughts?

```yaml
  preset:
    - name: Some preset
      default_target_temperature_low: 20
      default_target_temperature_high: 24
      fan_mode: 'off'
      mode: heat_cool
  default_preset: Some preset
```

Sort of thing. Would consolidate a couple of old behaviours + introduce more functionality in one swoop.

Thoughts?

kbx81 replied:
I like it...that's more or less what I had in mind. 😄

MrMDavidson replied:
Okay, so we're good to defer this change for VNext+1? Do you know if there's a way of marking an item in the schema as supported-but-deprecated with a warning? Or is all up to the docs to communicate that change?

kbx81 replied:
Yes, I'm fine either way. It was just a thought...I'm sure someone will want that before long. 😄

There _is_ some way to throw a warning during validation -- I know I've seen it before but I can't recall where. I bet a quick search for "deprecated" in the repo will likely turn up a result/example of how to do that. 😁

kbx81 commented on esphome/components/thermostat/thermostat_climate.cpp:
```diff
@@ -910,30 +923,110 @@ bool ThermostatClimate::supplemental_heating_required_() {
+  // Note: The mode, fan_mode, and swing_mode can all be defined on the preset but if the climate.control call
+  // also specifies them then the control's version will override these for that call
+  if (config.mode_.has_value()) {
+    this->mode = *config.mode_;
```
You need to call `refresh()` at the end of this function or use `switch_to_mode_()` to update the component's internal `mode` variable. This ensures the the appropriate actions are called immediately upon the mode change.

kbx81 commented on esphome/components/thermostat/thermostat_climate.cpp:
```diff
@@ -910,30 +923,110 @@ bool ThermostatClimate::supplemental_heating_required_() {
+  }
+
+  if (config.fan_mode_.has_value()) {
+    this->fan_mode = *config.fan_mode_;
```
You need to call `refresh()` at the end of this function or use `switch_to_fan_mode_()` to update the component's internal `fan_mode` variable. This ensures the the appropriate actions are called immediately upon the fan_mode change.

kbx81 commented on esphome/components/thermostat/thermostat_climate.cpp:
```diff
@@ -910,30 +923,110 @@ bool ThermostatClimate::supplemental_heating_required_() {
+
+  if (config.swing_mode_.has_value()) {
+    ESP_LOGV(TAG, "Setting swing mode to %s", LOG_STR_ARG(climate::climate_swing_mode_to_string(*config.swing_mode_)));
+    this->swing_mode = *config.swing_mode_;
```
You need to call `refresh()` at the end of this function or use `switch_to_swing_mode_()` to update the component's internal `swing_mode` variable. This ensures the the appropriate actions are called immediately upon the swing mode change.

kbx81 commented on esphome/components/thermostat/thermostat_climate.cpp:
```diff
@@ -910,30 +923,110 @@ bool ThermostatClimate::supplemental_heating_required_() {
+  if (this->preset != preset) {
+    Trigger<> *trig = this->preset_change_trigger_;
+    assert(trig != nullptr);
+    trig->trigger();
```
You could add `refresh();` after this line (just outside of the `if`) in lieu of using the individual functions I mentioned above.

kbx81 reviewed (changes requested):
Sorry it took me a while -- left some comments regarding a few fixes that need to be made. Nothing critical but still important polish to work in before we merge this in.

Please ping me if you have any questions or need help! 😄

MrMDavidson committed:
Sort imports

MrMDavidson committed:
Improve configuration validation for presets
- Unify temperature validation across default, away, and preset configuration
- Validate modes for presets have the required actions

MrMDavidson committed:
Trigger a refresh after changing internals of the thermostat

MrMDavidson committed:
Apply formatting fixes

MrMDavidson committed:
Validate mode, fan_mode, and swing_mode on presets

MrMDavidson committed:
Add preset temperature validation against visual min/max configuration

MrMDavidson committed:
Apply code formatting fixes

MrMDavidson committed:
Fix preset temperature validation

kbx81 commented:
@MrMDavidson you good with this? anything else you want to tweak? 🙂

MrMDavidson commented:
> @MrMDavidson you good with this? anything else you want to tweak? 🙂

Nope - I'm happy with this being merged.

kbx81 commented:
Done 🎉

MrMDavidson commented:
Yay! Thanks for your help getting this across the line!

kbx81 commented:
Of course -- and thank you for working so diligently on it and working all the suggestions and fixes. Super stoked to have this in place! 😄🍻
